File: Compressor.cpp
#include "CustomLib.hpp"

// C++ neural network model defined with value semantics
struct Net : torch::nn::Module
{
    Net(torch::Device device, int64_t n_feature, int64_t n_hidden1, int64_t n_hidden2, int64_t n_hidden3, int64_t n_output)
    {
        this->fc1     = register_module("fc1",     torch::nn::Linear(n_feature, n_hidden1));
        this->fc2     = register_module("fc2",     torch::nn::Linear(n_hidden1, n_hidden2));
        this->fc3     = register_module("fc3",     torch::nn::Linear(n_hidden2, n_hidden3));
        this->predict = register_module("predict", torch::nn::Linear(n_hidden3, n_output));

        this->fc1->to(device);
        this->fc2->to(device);
        this->fc3->to(device);
        this->predict->to(device);
    }

    torch::Tensor forward(torch::Tensor feature, bool getEncoded)
    {
        feature = torch::elu(this->fc1(feature));
        feature = torch::elu(this->fc2(feature));

        if (getEncoded)
            return feature;
    
        feature = torch::elu(this->fc3(feature));
        feature = this->predict(feature);
        return feature;
    }

    torch::nn::Linear fc1 = nullptr;
    torch::nn::Linear fc2 = nullptr;
    torch::nn::Linear fc3 = nullptr;
    torch::nn::Linear predict = nullptr;
};

using namespace std;

int main()
{
    auto start = chrono::high_resolution_clock::now();

    cout << "[Device: " << at::cuda::getDeviceProperties(0)->name << "]" << endl << endl;

    // VARIABLES
    const int64_t n_feature = 539;
    const int64_t n_hidden1 = 512;
    const int64_t n_hidden2 = 32;
    const int64_t n_hidden3 = 512;
    const int64_t n_output = 539;
    const int64_t batch_size = 32;

    // INIT NN
    auto model = Net(device, n_feature, n_hidden1, n_hidden2, n_hidden3, n_output);
    model.to(device);

    // DATASET
    vector<torch::Tensor> x, xNorm;
    LoadData("Poses", x, false);

	// NORMALIZE DATA
    for (int i = 0; i < (int) x.size(); i++)
    {
        auto mean = x[i].mean(1, true);
        auto std  = x[i].std(1, true).reshape({x[i].size(0), 1});

        xNorm.push_back((x[i] - mean) / std);
    }

    auto dataset = CustomDataset(xNorm, xNorm, 1, 0).map(torch::data::transforms::Stack<>());

    // DATA LOADER
    auto data_loader = torch::data::make_data_loader<torch::data::samplers::RandomSampler>
    (std::move(dataset), torch::data::DataLoaderOptions().batch_size(batch_size).workers(4));

    // OPTIMIZER
    torch::optim::Adam optimizer(model.parameters(), 0.001);
	torch::optim::StepLR scheduler(optimizer, 100, 0.99);

    std::cout << "Training Compressor...\n";

    for (size_t epoch = 1; epoch <= 50000; epoch++)
    {
        int batchId = 0;
        float epochLoss = 0;

        // Iterate the data loader to yield batches from the dataset.
        for (auto &batch : *data_loader)
        {
            batch.data = batch.data.to(device);
            batch.target = batch.target.to(device);

            // Execute the model on the input data.
            torch::Tensor prediction = model.forward(batch.data, false);
            
            // Compute a loss value to judge the prediction of our model.
            torch::Tensor loss = torch::nn::MSELoss()(prediction, batch.target);

            // Reset gradients.
            optimizer.zero_grad();

            // Compute gradients of the loss w.r.t. the parameters of our model.
            loss.backward();

            // Update the parameters based on the calculated gradients.
            optimizer.step();
            scheduler.step();
            
            epochLoss += prediction.size(0) * loss.item<float>();
            
            // Output the loss and checkpoint every 100 batches.
            // if (++batchId % 9 == 0 && epoch % 5000 == 0)
            // {
            //     std::cout << "Epoch: " << epoch << " | Loss: " << loss.item<float>() << std::endl;
            //     // Serialize your model periodically as a checkpoint.
            //     // torch::save(model, "Compressor.pt");
            // }
        }

        if (epoch % 5000 == 0)
        {
            cout << "Epoch: " << epoch << " | Loss: " << epochLoss << endl;
        }
    }

    // SET PRECISION
    // std::cout << std::fixed << std::setprecision(4);

    std::cout << "Finished\n";

	// SAVE WEIGHTS
	ofstream weights("/home/pau1o-hs/Documents/NNWeights/Compressor.txt");
    ofstream latents("/home/pau1o-hs/Documents/Database/LatentVariables.txt");
	torch::Tensor weightData;

	// FC1
	weightData = model.fc1->weight.cpu().detach().transpose(0, 1).flatten();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	weightData = model.fc1->bias.cpu().detach();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	// FC2
	weightData = model.fc2->weight.cpu().detach().transpose(0, 1).flatten();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	weightData = model.fc2->bias.cpu().detach();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

    // FC3
	weightData = model.fc3->weight.cpu().detach().transpose(0, 1).flatten();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	weightData = model.fc3->bias.cpu().detach();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	// PREDICT
	weightData = model.predict->weight.cpu().detach().transpose(0, 1).flatten();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	weightData = model.predict->bias.cpu().detach();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;


    for (int k = 0; k < (int) xNorm.size(); k++)
    {
        torch::Tensor encode = model.forward(xNorm[k], true);
        cout << encode.size(0) << ' ' << encode.size(1) << endl;

        for (int i = 0; i < (int) encode.size(0); i++)
        {
            for (int j = 0; j < (int) encode.size(1); j++)
                latents << encode[i][j].item<float>() << ' ';
            
            latents << endl;
        }
    }

    // RUNTIME
    auto stop = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::seconds>(stop - start);
    cout << "Runtime: " << (float) (duration.count() / 60.0f) << " minutes" << endl;
}

File: Compressor.py
import torch
from torch.autograd import Variable
import torch.nn.functional as F
import torch.utils.data as Data
import torch_optimizer as optim
from torch.utils.tensorboard import SummaryWriter

import numpy as np
import time

start_time = time.time()

compressorIn = []

with open("/home/pau1o-hs/Documents/Database/YData.txt") as f:
    for line in f:
        inner_list = [elt.strip() for elt in line.split(' ')]

        if inner_list == ['']:
            continue

        converted =  np.asarray(inner_list, dtype=np.float64, order='C')
        compressorIn.append(converted)

print(torch.cuda.get_device_name(0)) 
device = torch.device("cuda")
# device = torch.device("cpu")

x = torch.tensor(compressorIn, dtype=torch.float).to(device)        # x data (tensor), shape=(100, 1)
y = torch.tensor(compressorIn, dtype=torch.float).to(device)
print(x.size(1))

# torch can only train on Variable, so convert them to Variable
# x, y = Variable(x), Variable(y)

means = x.mean(dim=1, keepdim=True)
stds  = x.std(dim=1, keepdim=True)
normalized_data = (x - means) / stds

# print(means)
# print(means.shape)
# print(stds)
# print(stds.shape)
# print(normalized_data)

train = Data.TensorDataset(normalized_data, normalized_data)
train_loader = Data.DataLoader(train, shuffle=True, batch_size=32)

class Net(torch.nn.Module):
    def __init__(self, n_feature, n_hidden, n_hidden2, n_hidden3, n_output):
        super(Net, self).__init__()
        self.hidden = torch.nn.Linear(n_feature, n_hidden)   # hidden layer
        self.hidden2 = torch.nn.Linear(n_hidden, n_hidden2)   # hidden layer
        self.hidden3 = torch.nn.Linear(n_hidden2, n_hidden3)   # hidden layer
        self.predict = torch.nn.Linear(n_hidden3, n_output)   # output layer

    def forward(self, x, getEncoded):
        x = F.elu(self.hidden(x))      # activation function for hidden layer
        x = F.elu(self.hidden2(x))      # activation function for hidden layer

        if (getEncoded):
            return x

        x = F.elu(self.hidden3(x))      # activation function for hidden layer
        x = self.predict(x)             # linear output
        return x

net = Net(n_feature=1232, n_hidden=512, n_hidden2=32, n_hidden3=512, n_output=1232).to(device)     # define the network

optimizer = optim.RAdam(net.parameters(), lr=0.001)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1000, gamma=0.99)
loss_func = torch.nn.MSELoss()  # this is for regression mean squared loss

writer = SummaryWriter()

# train the network
for t in range(20000):
    epochLoss = 0.0

    for batch_idx, (data, target) in enumerate(train_loader):
        
        data.to(device), target.to(device)
        
        prediction = net(data, False)     # input x and predict based on x
        
        loss = loss_func(prediction, target)     # must be (1. nn output, 2. target)
        
        optimizer.zero_grad()   # clear gradients for next train
        loss.backward()         # backpropagation, compute gradients
        optimizer.step()        # apply gradients
        scheduler.step()

        epochLoss += loss * prediction.size(0)

    if t % 500 == 0:
        print(t, epochLoss.item())
    
    writer.add_scalar('Python Compressor Loss', epochLoss, t)


torch.set_printoptions(precision=6)
np.set_printoptions(precision=7, floatmode='fixed', suppress=True)

print("Runtime: %s minutes" % ((time.time() - start_time) / 60))

# prediction = net(normalized_data[0])

# print(x[0].cpu().detach().numpy())
# print()
# print(normalized_data[0].cpu().detach().numpy())
# print()
# print(prediction.cpu().detach().numpy())

with open('/home/pau1o-hs/Documents/NNWeights/Compressor.txt', "w+") as f:
    np.savetxt(f, net.hidden.weight.cpu().detach().numpy().transpose(), delimiter="\n")    
    np.savetxt(f, net.hidden.bias.cpu().detach().numpy(), delimiter="\n")

    np.savetxt(f, net.hidden2.weight.cpu().detach().numpy().transpose(), delimiter="\n")    
    np.savetxt(f, net.hidden2.bias.cpu().detach().numpy(), delimiter="\n")

    np.savetxt(f, net.hidden3.weight.cpu().detach().numpy().transpose(), delimiter="\n")    
    np.savetxt(f, net.hidden3.bias.cpu().detach().numpy(), delimiter="\n")

    np.savetxt(f, net.predict.weight.cpu().detach().numpy().transpose(), delimiter="\n")
    np.savetxt(f, net.predict.bias.cpu().detach().numpy(), delimiter="\n")

with open('/home/pau1o-hs/Documents/Database/LatentVariables.txt', "w+") as f:
    for i in range(len(normalized_data)):
        prediction = net(normalized_data[i], True)
        np.savetxt(f, prediction.cpu().detach().numpy()[None], delimiter=" ")


